Add support for isDelegatedRequest to ApplePayRequestBase

WDUK opened the pull request:
#### 3508d9e5a87243218e4181ec6cef677a29eff039
<pre>
Add support for isDelegatedRequest to ApplePayRequestBase
<a href="https://bugs.webkit.org/show_bug.cgi?id=303631">https://bugs.webkit.org/show_bug.cgi?id=303631</a>
<a href="https://rdar.apple.com/163791929">rdar://163791929</a>

Reviewed by Abrar Rahman Protyasha.

This change adds support for the isDelegatedRequest property to ApplePayRequestBase, allowing web content to indicate whether an Apple Pay request is delegated. The property is exposed through the Payment Request API and propagated through to PassKit.

The implementation includes:

* LayoutTests/http/tests/paymentrequest/paymentrequest-isDelegatedRequest.https-expected.txt: Added.
* LayoutTests/http/tests/paymentrequest/paymentrequest-isDelegatedRequest.https.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
Add new test and skip on platforms where it&apos;s not supported.

* Source/WTF/wtf/PlatformEnableCocoa.h:
Add ENABLE_APPLE_PAY_DELEGATED_REQUEST feature flag.

* Source/WebCore/Modules/applepay/ApplePayFeature.h:
* Source/WebCore/Modules/applepay/ApplePayFeature.idl:
Add PaymentRequestDelegation feature.

* Source/WebCore/Modules/applepay/ApplePayRequestBase.cpp:
(WebCore::convertAndValidate):
* Source/WebCore/Modules/applepay/ApplePayRequestBase.h:
* Source/WebCore/Modules/applepay/ApplePayRequestBase.idl:
Add isDelegatedRequest property.

* Source/WebCore/Modules/applepay/ApplePaySessionPaymentRequest.h:
Add isDelegatedRequest getter and setter, update constructor.

* Source/WebCore/PAL/pal/spi/cocoa/PassKitSPI.h:
Add SPI declaration for PKPaymentRequest isDelegatedRequest property.

* Source/WebCore/testing/MockPaymentCoordinator.cpp:
(WebCore::MockPaymentCoordinator::showPaymentUI):
* Source/WebCore/testing/MockPaymentCoordinator.h:
(WebCore::MockPaymentCoordinator::isDelegatedRequest const):
* Source/WebCore/testing/MockPaymentCoordinator.idl:
Add isDelegatedRequest support to mock payment coordinator for testing.

* Source/WebKit/Configurations/AllowedSPI.toml:
Add SPI staging entry with cleanup rdar.

* Source/WebKit/Shared/ApplePay/cocoa/WebPaymentCoordinatorProxyCocoa.mm:
(WebKit::toPKPaymentRequest):
Propagate isDelegatedRequest to PKPaymentRequest.

* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
Add isDelegatedRequest to serialization.

Canonical link: <a href="https://commits.webkit.org/304173@main">https://commits.webkit.org/304173@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/499e4099012877e334f8216aee406557bb6473d5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134354 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6828 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45570 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141898 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86355 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/136224 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7421 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6694 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102701 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/69960 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137301 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5174 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120409 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83492 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5039 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2657 "Passed tests") | [⏳ 🛠 wpe-cairo-libwebrtc ](https://ews-build.webkit.org/#/builders/WPE-Cairo-LibWebRTC-Build-EWS "Waiting in queue, processing has not started yet") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/126415 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/114245 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38550 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144578 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/132864 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6507 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39133 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/111100 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6584 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5467 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111367 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28339 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4872 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116685 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/60289 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6559 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34885 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/165800 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6386 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70124 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43329 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6621 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6498 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/Modules/applepay/ApplePaySessionPaymentRequest.h:
```diff
@@ -276,6 +284,9 @@ class ApplePaySessionPaymentRequest {
             , m_merchantCategoryCode(WTFMove(merchantCategoryCode))
+#endif
+#if ENABLE(APPLE_PAY_DELEGATED_REQUEST)
+            , m_isDelegatedRequest(WTFMove(isDelegatedRequest))
```
The `WTFMove()` is unnecessary since you are only passing in an lvalue

aprotyas commented on Source/WebCore/Modules/applepay/ApplePaySessionPaymentRequest.h:
```diff
@@ -184,6 +184,11 @@ class ApplePaySessionPaymentRequest {
 #endif
 
+#if ENABLE(APPLE_PAY_DELEGATED_REQUEST)
+    const std::optional<bool> isDelegatedRequest() const { return m_isDelegatedRequest; }
```
A const lvalue return type is not super helpful, it should either return a reference or just a plain regular copy.


```suggestion
   std::optional<bool> isDelegatedRequest() const { return m_isDelegatedRequest; }
```

or

```suggestion
    const std::optional<bool> & isDelegatedRequest() const { return m_isDelegatedRequest; }
```

I would go with the former.

aprotyas reviewed (commented):
everything looks good, one editorial change.

```
    Add support for isDelegatedRequest to ApplePayRequestBase

    https://bugs.webkit.org/show_bug.cgi?id=303631
    rdar://163791929
```

^ do you mind removing the extra line between the commit title and the bugzilla link

marcoscaceres commented on LayoutTests/http/tests/paymentrequest/paymentrequest-isDelegatedRequest.https.html:
```diff
@@ -0,0 +1,62 @@
+<script>
+    setup({ explicit_done: true, explicit_timeout: true });
+    test((test) => {
+        const method = validPaymentMethod();
```
Nothing to fix here... but as a follow up, we should really allow `validPaymentMethod` to take a features argument or something (or as data object to spread over the valid payment method).

marcoscaceres commented on LayoutTests/http/tests/paymentrequest/paymentrequest-isDelegatedRequest.https.html:
```diff
@@ -0,0 +1,62 @@
+        
+        const response = await request.show();
+        
+        assert_equals(internals.mockPaymentCoordinator.isDelegatedRequest, method.data.isDelegatedRequest, "check that `isDelegatedRequest` still matches after the payment is accepted");
```
Nit: line length.

```suggestion
        assert_equals(
            internals.mockPaymentCoordinator.isDelegatedRequest,
            method.data.isDelegatedRequest,
            "check that `isDelegatedRequest` still matches after the payment is accepted"
        );
```

marcoscaceres reviewed (approved):
Just a nit... but feel free to ignore if you just want to merge.

Commit:
Add support for isDelegatedRequest to ApplePayRequestBase
https://bugs.webkit.org/show_bug.cgi?id=303631
rdar://163791929

Reviewed by Abrar Rahman Protyasha.

This change adds support for the isDelegatedRequest property to ApplePayRequestBase, allowing web content to indicate whether an Apple Pay request is delegated. The property is exposed through the Payment Request API and propagated through to PassKit.

The implementation includes:

* LayoutTests/http/tests/paymentrequest/paymentrequest-isDelegatedRequest.https-expected.txt: Added.
* LayoutTests/http/tests/paymentrequest/paymentrequest-isDelegatedRequest.https.html: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk2/TestExpectations:
Add new test and skip on platforms where it's not supported.

* Source/WTF/wtf/PlatformEnableCocoa.h:
Add ENABLE_APPLE_PAY_DELEGATED_REQUEST feature flag.

* Source/WebCore/Modules/applepay/ApplePayFeature.h:
* Source/WebCore/Modules/applepay/ApplePayFeature.idl:
Add PaymentRequestDelegation feature.

* Source/WebCore/Modules/applepay/ApplePayRequestBase.cpp:
(WebCore::convertAndValidate):
* Source/WebCore/Modules/applepay/ApplePayRequestBase.h:
* Source/WebCore/Modules/applepay/ApplePayRequestBase.idl:
Add isDelegatedRequest property.

* Source/WebCore/Modules/applepay/ApplePaySessionPaymentRequest.h:
Add isDelegatedRequest getter and setter, update constructor.

* Source/WebCore/PAL/pal/spi/cocoa/PassKitSPI.h:
Add SPI declaration for PKPaymentRequest isDelegatedRequest property.

* Source/WebCore/testing/MockPaymentCoordinator.cpp:
(WebCore::MockPaymentCoordinator::showPaymentUI):
* Source/WebCore/testing/MockPaymentCoordinator.h:
(WebCore::MockPaymentCoordinator::isDelegatedRequest const):
* Source/WebCore/testing/MockPaymentCoordinator.idl:
Add isDelegatedRequest support to mock payment coordinator for testing.

* Source/WebKit/Configurations/AllowedSPI.toml:
Add SPI staging entry with cleanup rdar.

* Source/WebKit/Shared/ApplePay/cocoa/WebPaymentCoordinatorProxyCocoa.mm:
(WebKit::toPKPaymentRequest):
Propagate isDelegatedRequest to PKPaymentRequest.

* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
Add isDelegatedRequest to serialization.

Canonical link: https://commits.webkit.org/304173@main
